src/core/ext/lb_policy/round_robin/round_robin.c

```diff
@@ -31,6 +31,34 @@
  *
  */
 
+/** Round Robin Policy.
+ *
+ * This policy keeps:
+ * - A circular list of ready (connected) subchannels, the
+ *   *readylist*. An empty readylist consists solely of its root (dummy) node.
+ * - A pointer to the last element picked from the readylist, the *lastpick*.
+ *   Initially set to point to the readylist's root.
+ *
+ * Behavior:
+ * - When a subchannel connects, it's *prepended* to the readylist's root node.
+ *   Ie, if readylist = A <-> B <-> ROOT <-> C
+ *                      ^                    ^
+ *                      |____________________|
+ *   and subchannel D becomes connected, the addition of D to the readylist
+ *   results in  readylist = A <-> B <-> D <-> ROOT <-> C
+ *                           ^                          ^
+ *                           |__________________________|
+ * - When a subchannel disconnects, it's removed from the readylist. If the
+ *   subchannel being removed was the most recently picked, the *lastpick*
+ *   pointer moves to the removed node's previous element. Note that if the
+ *   readylist only had one element, this is still legal, as the lastpick would
+ *   point to the dummy root node, for an empty readylist.
+ * - Upon picking, *lastpick* is updated to point to the returned (connected)
+ *   subchannel. Note that it possible that the selected subchannel becomes
+ *   disconnected in the interim between the selection and the actual usage of
+ *   the subchannel by the caller.
```

Comment:
@zhangkun83 I think we can have more than one variations of round robin in grpc. I think what David described here is the most common and traditional one and should satisfy most of users. We can definitely have another which implements the policy you proposed here. We just need to make sure the interface is flexible enough to support all these variations (it seems your proposal only needs very limited API since it does not need to interact with grpc internals so that I have some slight concern about java API. If it is not the case, just ignore this.).

BTW, failing the RPC might not be correct decision. I think the right decision should depend on whether it is fail-fast.